File: packages/cli-py/planforge/utils/paths.py
"""Path resolution for PlanForge (project root, .cursor/plans, .cursor/contexts)."""

import os
from datetime import datetime
from pathlib import Path


def get_project_root(cwd: str | None = None) -> str:
    cwd = cwd or os.getcwd()
    dir_path = Path(cwd).resolve()
    while True:
        if (
            (dir_path / "planforge.json").exists()
            or (dir_path / ".cursor" / "plans").exists()
            or (dir_path / ".cursor" / "contexts").exists()
        ):
            return str(dir_path)
        parent = dir_path.parent
        if parent == dir_path:
            break
        dir_path = parent
    return cwd


def get_cursor_dir(project_root: str) -> str:
    return str(Path(project_root) / ".cursor")


def get_plans_dir(project_root: str) -> str:
    return str(Path(project_root) / ".cursor" / "plans")


def get_contexts_dir(project_root: str) -> str:
    return str(Path(project_root) / ".cursor" / "contexts")


def get_context_dir(project_root: str) -> str:
    return get_contexts_dir(project_root)


def get_default_context_dirs(project_root: str) -> list[str]:
    return [get_contexts_dir(project_root)]


def get_date_parts(date: datetime | None = None) -> tuple[str, str, str]:
    """Return (yyyy_mm_dd, mmdd, hhmm)."""
    now = date or datetime.now()
    return now.strftime("%Y-%m-%d"), now.strftime("%m%d"), now.strftime("%H%M")


def get_dated_plans_dir(project_root: str, date: datetime | None = None) -> str:
    yyyy_mm_dd, _, _ = get_date_parts(date)
    return str(Path(get_plans_dir(project_root)) / yyyy_mm_dd)


def get_templates_root() -> str:
    """Resolve repo templates dir (monorepo: packages/cli-py/planforge/utils -> repo/templates)."""
    base = Path(__file__).resolve().parent.parent.parent.parent.parent
    return str(base / "templates")


def get_prompts_dir() -> str:
    """Resolve prompts directory from installed planforge_core package."""
    import planforge_core
    return str(Path(planforge_core.__file__).resolve().parent.parent / "prompts")


def get_models_json_path() -> str:
    """Resolve models.json path from installed planforge_core package."""
    import planforge_core
    return str(Path(planforge_core.__file__).resolve().parent.parent / "models.json")

File: packages/cli-js/src/utils/paths.ts
/**
 * Path resolution for PlanForge (project root, .cursor/plans, .cursor/contexts)
 */

import { existsSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));

/**
 * Find project root by walking up from cwd until we find planforge.json or .cursor/plans or .cursor/contexts.
 */
export function getProjectRoot(cwd: string = process.cwd()): string {
  let dir = resolve(cwd);
  const root = resolve(dir, "..");
  while (dir !== root) {
    if (
      existsSync(resolve(dir, "planforge.json")) ||
      existsSync(resolve(dir, ".cursor", "plans")) ||
      existsSync(resolve(dir, ".cursor", "contexts"))
    ) {
      return dir;
    }
    dir = resolve(dir, "..");
  }
  return cwd;
}

export function getPlansDir(projectRoot: string): string {
  return resolve(projectRoot, ".cursor", "plans");
}

export function getContextsDir(projectRoot: string): string {
  return resolve(projectRoot, ".cursor", "contexts");
}

export function getContextDir(projectRoot: string): string {
  return getContextsDir(projectRoot);
}

export function getDefaultContextDirs(projectRoot: string): string[] {
  return [getContextsDir(projectRoot)];
}

export function getDateParts(date: Date = new Date()): { yyyyMmDd: string; mmdd: string; hhmm: string } {
  const year = String(date.getFullYear());
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const day = String(date.getDate()).padStart(2, "0");
  const hours = String(date.getHours()).padStart(2, "0");
  const minutes = String(date.getMinutes()).padStart(2, "0");
  return {
    yyyyMmDd: `${year}-${month}-${day}`,
    mmdd: `${month}${day}`,
    hhmm: `${hours}${minutes}`,
  };
}

export function getDatedPlansDir(projectRoot: string, date: Date = new Date()): string {
  return resolve(getPlansDir(projectRoot), getDateParts(date).yyyyMmDd);
}

/**
 * Resolve path relative to CLI package (for templates).
 * From dist/utils/paths.js: two levels up = package root; templates live next to dist/ when published.
 */
export function getTemplatesRoot(): string {
  return resolve(__dirname, "..", "..", "templates");
}

/**
 * Resolve core root (prompts, models.json). Always uses the bundled copy at package root; no @planforge/core dependency.
 */
function getCoreRoot(): string {
  return resolve(__dirname, "..", "..", "core");
}

/**
 * Resolve prompts directory (planner/implementer system prompts). Uses bundled core from build.
 */
export function getPromptsDir(): string {
  return resolve(getCoreRoot(), "prompts");
}

/**
 * Resolve models.json path (planforge model command). Uses bundled core from build.
 */
export function getModelsJsonPath(): string {
  return resolve(getCoreRoot(), "models.json");
}
